### PR TITLE
Fix problem with scaling general loop blocks; add dumpers

### PR DESCRIPTION
**Pull request by BruceForstall**
1. Fix a problem where scaling block weights of a range of basic blocks that compose a general loop could encounter a block that is unreachable. This can happen for various kinds of blocks that the JIT doesn't like to remove even if unreachable. Simply skip scaling those blocks.

2. Add dumpers for `FlowGraphDfsTree` and `BlockToNaturalLoopMap`. These can be called from the debugger (perhaps they should be called to output these things to the JitDump).

3. Update `fgDfsBlocksAndRemove()` to dump any blocks that it did not remove, even if they were unreachable.

This was found as part of fixing JitOptRepeat: https://github.com/dotnet/runtime/pull/94250

### CODE REVIEW

**Comment by BruceForstall** [comment]
@jakobbotsch PTAL
cc @dotnet/jit-contrib

File: src/coreclr/jit/flowgraph.cpp [hunk 1]
```diff
@@ -6131,6 +6151,37 @@ BlockToNaturalLoopMap* BlockToNaturalLoopMap::Build(FlowGraphNaturalLoops* loops
...
+        {
+            if (m_indices[i] == UINT_MAX)
+            {
+                // Just leave the loop space empty if there is no enclosing loop
+                printf("%02u : \n", i);
+            }
+            else
+            {
+                printf("%02u : %02u\n", i, m_indices[i]);
+            }
```
Comment (jakobbotsch) [inline src/coreclr/jit/flowgraph.cpp:6178]:
```suggestion
            if (m_indices[i] == UINT_MAX)
            {
                // Just leave the loop space empty if there is no enclosing loop
                printf(FMT_BB " : \n", dfs->GetPostOrder(i)->bbNum);
            }
            else
            {
                printf(FMT_BB " : " FMT_LP "\n", dfs->GetPostOrder(i)->bbNum, m_indices[i]);
            }
```

File: src/coreclr/jit/fgopt.cpp [hunk 2]
```diff
@@ -5309,6 +5309,24 @@ PhaseStatus Compiler::fgDfsBlocksAndRemove()
...
+                for (BasicBlock* block : Blocks())
+                {
+                    if (!m_dfsTree->Contains(block))
+                    {
+                        printf("  " FMT_BB "\n", block->bbNum);
+                    }
+                }
+            }
+        }
+#endif // DEBUG
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgopt.cpp:5328]: Luckily the blocks are still removed in the "graph" sense -- they don't have any preds/succs left anymore, so they don't interact with reachable blocks in any way. I.e. asserts like these are ok:
https://github.com/dotnet/runtime/blob/b60a541554cc075b903a4d218456ef8588092a8c/src/coreclr/jit/morph.cpp#L13833-L13835

But yeah, we still have this annoying case to check for when iterating the block list.

File: src/coreclr/jit/flowgraph.cpp [hunk 3]
```diff
@@ -3907,6 +3907,26 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
...
+void FlowGraphDfsTree::Dump() const
+{
+    printf("DFS tree. %s.\n", HasCycle() ? "Has cycle" : "No cycle");
+    printf("PO RPO -> BB [pre, post]\n");
+    for (unsigned i = 0; i < GetPostOrderCount(); i++)
+    {
+        unsigned          rpoNum = GetPostOrderCount() - i - 1;
+        BasicBlock* const block  = GetPostOrder(i);
+        printf("%02u %02u -> " FMT_BB "[%u, %u]\n", i, rpoNum, block->bbNum, block->bbPreorderNum,
+               block->bbPostorderNum);
```
[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/flowgraph.cpp:3924]: Well, this is a dumper, so this can help verify that in the output.

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**